src/tpm2_createpolicy.c

```diff
@@ -0,0 +1,330 @@
+        return -1;
+    }
+
+    // Need to flush the session here.
+    rval = Tss2_Sys_FlushContext( ctx->sapi_context,
+        ctx->policy_session->sessionHandle );
+    if( rval!=TPM_RC_SUCCESS ){
+        return rval;
+    }
+
+    // And remove the session from sessions table.
+    rval = tpm_session_auth_end(ctx->policy_session);
+    if( rval!=TPM_RC_SUCCESS ){
+        return rval;
+    }
+
+    return rval;
+}
+
+static bool init(int argc, char *argv[], char *envp[], createpolicypcr_ctx *ctx) {
+    
+    unsigned policyfile_flag=0, pcr_index_flag=0, policy_type_pcr_flag=0,
+             pcr_alg_flag=0; 
+    
+    struct option sOpts[] = {
+            { "policyfile",     required_argument,  NULL,  'F' },
+            { "policyPCR",      no_argument,        NULL,  'P' },
+            { "pcrIndex",       required_argument,  NULL,  'I' },
+            { "pcrAlg",         required_argument,  NULL,  'A' },
+            { "pcrInputFile",   required_argument,  NULL,  'Q' },
+            { NULL,             no_argument,        NULL, '\0' },
+    };
+
+    if (argc == 1) {
+        execute_man(argv[0], envp);
+        return false;
+    }
+
+    if (argc > (int) (2 * sizeof(sOpts) / sizeof(struct option))) {
+        showArgMismatch(argv[0]);
+        return false;
+    }
+
+    int opt;
+    bool result;
+    while ((opt = getopt_long(argc, argv, "PF:I:A:Q:", sOpts, NULL)) != -1) {
+        switch (opt) {
+        case 'F':
+            snprintf(ctx->policyfile, sizeof(ctx->policyfile), "%s", optarg);
+            policyfile_flag=1;
+            break;
+        case 'P':
+            printf("Policy type chosen is policyPCR.\n");
+            policy_type_pcr_flag = 1;
+            break;       
+        case 'I':
+            if(string_bytes_get_uint32(optarg, &ctx->pcr_index)!=true) {
+                return false;
+            }
+            if(ctx->pcr_index > 23 || ctx->pcr_index<0) {
+                printf("Invalid pcr_index %d. Choose between 0..23\n",
+                    ctx->pcr_index );
+                return false;
+            }
+            pcr_index_flag=1;
+            break;
+        case 'Q':
+            snprintf(ctx->raw_pcr_file, sizeof(ctx->raw_pcr_file), "%s", optarg);
+            ctx->raw_pcr_flag = 1;
+            break;
+        case 'A':
+            if(string_bytes_get_uint16(optarg, &ctx->hash_alg)!=true) {
+                return false;
+            }
+            pcr_alg_flag = 1;
+            break;
+        case '?':
+        default:
+            showArgMismatch(argv[0]);
+            return false;
+        }
+    }
+
+    if(!policyfile_flag || !pcr_index_flag) {
+        showArgMismatch(argv[0]);
+    }
+
+    if(policyfile_flag && pcr_index_flag) {
+        printf("Policy File = %s\n", ctx->policyfile );
+        printf("PCR Index= %d\n",ctx->pcr_index);
+        if( build_policy(ctx, build_pcr_policy) != TPM_RC_SUCCESS) {
```

Comment:
I don't like hiding functions with sideffects in conditional statements. This isn't in the coding standard yet, but i'd prefer this and others to be

retval = build_policy();
if (retval) ....